### PR TITLE
Add a tagged union type that replaces tensor in the interpreter.

### PR DESCRIPTION
**Pull request by zdevito**
IValue is short for interpreter value. It is used frequently so a short name is important.
This will allow us to implement more non-tensor types in an efficient way and remove
many hacks from the compiler.

This PR is limited. It only introduces IValue and changes interpreter to use it.
Follow up PRs will:
* Change the way aten_ops consume non-tensor types so that integer lists,
  are no longer represented as Tensors.
* Introduce TensorList as a fundamental type and remove all vararg handling in gen_jit_dispatch
* Change the compiler to implement math on primitive numbers rather than converting to tensors.

@jamesr66a  @apaszke

### CODE REVIEW

File: tools/jit/templates/register_aten_ops.cpp [hunk 1]
```diff
@@ -39,10 +38,16 @@ namespace {
...
     return -1;
-  auto & t = *(stack.end() - N);
+  auto t = (stack.end() - N)->toTensor();
   return t.type().is_cuda() ? (int) t.get_device() : -1;
 }
 
+std::vector<at::Tensor> toTensors(at::ArrayRef<IValue> ivalues) {
+  return fmap(ivalues, [](const IValue& v) {
+    return v.toTensor();
+  });
```
Comment (apaszke) [inline tools/jit/templates/register_aten_ops.cpp]: If we drop this into `functional.h`
```cpp
// Overload for a method without any arguments                                                      
template<typename F, typename T>                                                                    
inline auto fmap(const T& inputs, const F& fn) -> std::vector<decltype(((*inputs.begin()).*fn)())> {
  std::vector<decltype(((*inputs.begin()).*fn)())> r;                                               
  r.reserve(inputs.size());                                                                         
  for(const auto & input : inputs)                                                                  
    r.push_back((input.*fn)());                                                                     
  return r;                                                                                         
}                                                                                                   
```
then we'll be able to just do
```cpp
return fmap(ivalues, &IValue::toTensor);
```

File: torch/csrc/jit/graph_executor.cpp [hunk 2]
```diff
@@ -72,6 +73,17 @@ struct ExecutionPlanAutogradFunction : public autograd::Function {
 };
 
 
+// helper to run interpreter on variables until we switch
+// everything to IValue
+inline variable_tensor_list runOneStage(const Code & code, variable_tensor_list inputs) {
```
Comment (apaszke) [inline torch/csrc/jit/graph_executor.cpp]: I guess now that we don't have stages we could just call it `runWithVariables`

File: torch/csrc/jit/graph_executor.cpp [hunk 3]
```diff
@@ -72,6 +73,17 @@ struct ExecutionPlanAutogradFunction : public autograd::Function {
...
+  std::vector<IValue> stack(inputs.begin(), inputs.end());
+  InterpreterState(code).runOneStage(stack);
+  // note: we never unwrapped inputs, because we want autograd to record the trace
```
Comment (apaszke) [inline torch/csrc/jit/graph_executor.cpp]: Unwrapped as in not used Variables?

File: torch/csrc/jit/graph_executor.cpp [hunk 4]
```diff
@@ -114,11 +125,10 @@ struct ExecutionPlan {
...
   // inplace to avoid allocations
-  variable_tensor_list unwrapVariables(variable_tensor_list && list) const {
-    for(auto & v : list) {
-      v = v.defined() ? autograd::as_variable_ref(v).detach() : at::Tensor();
-    }
-    return std::move(list);
+  std::vector<IValue> unwrapVariables(variable_tensor_list && list) const {
+    return fmap(list, [](const Variable& v) -> IValue {
+      return v.defined() ? autograd::as_variable_ref(v).detach() : at::Tensor();
+    });
```
Comment (apaszke) [inline torch/csrc/jit/graph_executor.cpp]: 1. The comment above is no longer true.
2. Why do we take an r-value reference here? a const ref would work perfectly fine if we're reallocating anyway

File: torch/csrc/jit/ivalue.h [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+  }
+  PointerType* operator->() const {
+    return get();
+  }
+  operator bool() const {
+    return pImpl != nullptr;
+  }
+private:
+  PointerType * pImpl;
+};
```
Comment (apaszke) [inline torch/csrc/jit/ivalue.h]: This looks pretty much like a clone of `at::TensorBase`. Why can't we use that one (possibly after updating to use CRTP, which would actually let us remove some code from the `Tensor` class).

File: torch/csrc/jit/ivalue.h [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+// The tag is currently 4 bytes to determine the type, and 1 byte
+// to mark whether that type is a subtype of at::Retainable and needs
+// retain/release calls.
+struct IValue {
+  IValue()
+  : IValue(static_cast<int64_t>(0)) {}
+  IValue(const IValue& rhs)
+      : as_int(rhs.as_int),
+        tag(rhs.tag),
+        retainable(rhs.retainable) {
```
Comment (apaszke) [inline torch/csrc/jit/ivalue.h]: nit: the formatting in here is quite inconsistent (member initialization list of this and previous constructor)

File: torch/csrc/jit/ivalue.h [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+
+// IValue is the generic tagged union used by the interpreter to hold
+// all value types.
+// It is a 16-byte object with an 8-byte payload and an 8-byte tag.
+// The tag is currently 4 bytes to determine the type, and 1 byte
+// to mark whether that type is a subtype of at::Retainable and needs
+// retain/release calls.
+struct IValue {
+  IValue()
+  : IValue(static_cast<int64_t>(0)) {}
```
Comment (apaszke) [inline torch/csrc/jit/ivalue.h]: Hmmm why do we default construct to an int-based IValue? Maybe we should have a concept of `None`, or prohibit default construction altogether.

File: torch/csrc/jit/ivalue.h (hunk 6 above)
Comment (apaszke) [reply torch/csrc/jit/ivalue.h]: Also, I'm not sure if always copying the `as_int` field is well defined by the standard if the union actually contains a different type 😊 Can you please add a comment saying that we don't need to copy other types because `as_int` is part of a union? Maybe we should add another member to the union called `payload` to avoid mentioning any specific type unnecessarily?

File: torch/csrc/jit/ivalue.h [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+  // While some of these accessors could be generated through templates,
+  // we prefer to write them manually for clarity
+
+  // Tensor
+  IValue(at::Tensor t)
+  : tag(Tag::Tensor), retainable(t.defined())  {
+    // note: the undefined tensor is not refcounted, so while it
+    // is tagged as a tensor, retainable is set to false.
+    as_tensor_impl = t.at::detail::TensorBase::detach();
+  }
```
Comment (apaszke) [inline torch/csrc/jit/ivalue.h]: It would be nice to have an rvalue based constructor too.

File: torch/csrc/jit/ivalue.h [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+    return as_double;
+  }
+
+  // Int
+  IValue(int64_t i)
+  : tag(Tag::Int), retainable(false) {
+    as_int = i;
+  }
+  // allow you to pass literals (3, 4) without ambiguity
+  IValue(int32_t i)
```
Comment (apaszke) [inline torch/csrc/jit/ivalue.h]: FWIW those literals are `int` literals, and it so only happens in your compiler that it's the same type as `int32_t`. IIRC it's not the case in some weird cases, even on x86_64 systems with 32-bit ints.

File: torch/csrc/jit/ivalue.h [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+    clearToZero();
+    return t;
+  }
+  template<typename T>
+  Shared<T> toRetainable() const {
+    return Shared<T>(static_cast<T*>(as_retainable), true);
+  }
+  void clearToZero() {
+    as_int = 0;
+    tag = Tag::Int;
```
Comment (apaszke) [inline torch/csrc/jit/ivalue.h]: ugh why

File: torch/csrc/jit/ivalue.h [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+    retainable = false;
+  }
+  enum class Tag : uint32_t {
```
Comment (apaszke) [inline torch/csrc/jit/ivalue.h]: Is there any reason why we want this to be 4-bytes? Are you trying to prevent the compiler from padding this thing to 12 bytes only?

File: torch/csrc/jit/ivalue.h [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+ private:
+  ConstantList(std::vector<Elem> elements_)
+  : elements_(std::move(elements_)) {}
+  std::vector<Elem> elements_;
+ public:
+  static Shared<ConstantList<Elem>> create(std::vector<Elem> elements_) {
+    return Shared<ConstantList<Elem>>(
+        new ConstantList<Elem>(std::move(elements_)), false);
+  }
+  at::ArrayRef<Elem> elements() const {
```
Comment (apaszke) [inline torch/csrc/jit/ivalue.h]: It's a bit funny that you can modify the elements of a `ConstantList`. I guess we only care about having a constant length, right?

File: torch/csrc/jit/register_prim_ops.cpp [hunk 13]
```diff
@@ -42,7 +42,10 @@ RegisterOperators reg({
             autograd::profiler::RecordFunction record("FusionGroup");
             std::vector<at::Tensor> toutputs;
             // TODO: have fusion_fn work off of a stack as well
-            fusion_fn->launch(last(stack, num_inputs), toutputs);
+            auto tinputs = fmap(last(stack, num_inputs), [](const IValue& v) {
+              return v.toTensor();
+            });
```
Comment (apaszke) [inline torch/csrc/jit/register_prim_ops.cpp]: It's a bit annoying the we have all those new allocations 😕

File: torch/csrc/jit/register_prim_ops.cpp [hunk 14]
```diff
@@ -69,28 +72,14 @@ RegisterOperators reg({
...
-        }),
-
     Operator(
         prim::Print,
         [](Node* node) {
           size_t num_inputs = node->inputs().size();
           return [num_inputs](Stack& stack) {
             bool first = true;
-            for (at::Tensor i : last(stack, num_inputs)) {
+            for (IValue i_ : last(stack, num_inputs)) {
```
Comment (apaszke) [inline torch/csrc/jit/register_prim_ops.cpp]: Is this really meant to copy the IValue (missing `&`)?

File: torch/csrc/jit/script/compiler.cpp [hunk 15]
```diff
@@ -297,7 +297,7 @@ static bool isNumberSubtype(const TypePtr& type) {
 }
 
 at::optional<std::vector<int64_t>> getIntListAttribute(at::optional<int32_t> N, Value* input) {
-  auto list = constant_as<at::IntList>(input);
+  auto list = constant_as<std::vector<int64_t>>(input);
   if(list)
     return std::vector<int64_t>(*list);
```
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: this could change to just `return *list`.

File: torch/csrc/jit/interpreter.h [hunk 16]
```diff
@@ -44,7 +45,7 @@ struct InterpreterState {
   // advance the interpreter state by running one stage. Returning the
   // outputs for that stage, suspending the computation.
   // Call this function again continues computation where it left off.
-  void runOneStage(std::vector<at::Tensor> & stack);
+  void runOneStage(std::vector<IValue> & stack);
```
Comment (apaszke) [inline torch/csrc/jit/interpreter.h]: Can we please use `Stack` in here?

File: torch/csrc/jit/test_jit.cpp [hunk 17]
```diff
@@ -898,13 +904,58 @@ void testControlFlow() {
...
+  Shared<IntList> foo = IntList::create({3, 4, 5});
+  JIT_ASSERT(foo->use_count() == 1);
+  IValue bar(foo);
+  JIT_ASSERT(foo->use_count() == 2);
+  auto baz = bar;
+  JIT_ASSERT(foo->use_count() == 3);
+  auto foo2 = std::move(bar);
+  JIT_ASSERT(foo->use_count() == 3);
+  JIT_ASSERT(foo2.isIntList());
+  JIT_ASSERT(bar.isInt());
```
Comment (apaszke) [inline torch/csrc/jit/test_jit.cpp]: Well that's exactly something I'd like to avoid. Even if we make undefined `IValue`s be ints, we shouldn't test this.

File: torch/csrc/jit/test_jit.cpp [hunk 18]
```diff
@@ -898,13 +904,58 @@ void testControlFlow() {
...
+  foo2 = IValue(4.0);
+  JIT_ASSERT(foo2.isDouble());
+  JIT_ASSERT(foo2.toDouble() == 4.0);
+  JIT_ASSERT(foo->use_count() == 2);
+  JIT_ASSERT(baz.toIntList()->elements().equals({3,4,5}));
+
+  auto move_it = std::move(baz).toIntList();
+  JIT_ASSERT(foo->use_count() == 2);
+  JIT_ASSERT(baz.isInt());
```
Comment (apaszke) [inline torch/csrc/jit/test_jit.cpp]: Same here

**Review by apaszke — APPROVED** [review]
LGTM. Some suggestions that might make the code nicer, but nothing big.

File: torch/csrc/jit/graph_executor.cpp (hunk 2 above)
Comment (zdevito) [reply torch/csrc/jit/graph_executor.cpp]: The interpreter can still compute stages, and it adds complexity to the it as well. We should add a task to remove this functionality and then rename this function.

File: torch/csrc/jit/graph_executor.cpp (hunk 3 above)
Comment (zdevito) [reply torch/csrc/jit/graph_executor.cpp]: Holdover from when we computed on Tensors, I removed it.

File: torch/csrc/jit/graph_executor.cpp (hunk 4 above)
Comment (zdevito) [reply torch/csrc/jit/graph_executor.cpp]: Updated the comment. I want to keep the r-value here, because I hope to eventually figure out a way to keep the types lined up enough to avoid the copies that this patch introduces. Thoughts:

1. Use ShortVector so most of time we don't get allocations. Downsize: the copies/moves are expensive
2. Switch Autograd to store IValue rather than Tensor. Downside: potential overhead in autograd (though maybe avoidable if we have autograd use unchecked versions of toTensor internally).
3. Use our own vector type that can safely move its data array from compatible types. Prereq: Tensor/IValue need to have the same in-memory representation, or at least the same size.

For now I am not going to do any of these since they will take time and it is not clear it is a critical pathway. We should follow up with @zou3519 when he figures out what the slowness is for LSTMs but so far it seems unrelated to the executor and exists inside of aten.

File: torch/csrc/jit/ivalue.h (hunk 5 above)
Comment (zdevito) [reply torch/csrc/jit/ivalue.h]: Eventually, yes, but it is not close enough to make the refactoring simple. In addition to needing CRTP, the way TensorBase requires different handling for Undefined adds additional complexity that this wrapper type doesn't need. We can probably figure out a way to de-dup but because it wasn't trivial I chose to just move on.

File: torch/csrc/jit/ivalue.h (hunk 6 above)
Comment (zdevito) [reply torch/csrc/jit/ivalue.h]: I did the payload thing. I think this is one of the strict aliasing issues. If we run into problems we can see if there is a way to coerce the compiler into generating the same code while still technically accessing the right field.

File: torch/csrc/jit/ivalue.h (hunk 7 above)
Comment (zdevito) [reply torch/csrc/jit/ivalue.h]: Ill add a None, it fixes the issues you have brought up.

File: torch/csrc/jit/ivalue.h (hunk 12 above)
Comment (zdevito) [reply torch/csrc/jit/ivalue.h]: ArrayRef is an immutable view as far as I understand, so I don't believe you can modify the members.

File: torch/csrc/jit/register_prim_ops.cpp (hunk 13 above)
Comment (zdevito) [reply torch/csrc/jit/register_prim_ops.cpp]: Agreed, and I outlined some options above.

File: torch/csrc/jit/ivalue.h (hunk 8 above)
Comment (apaszke) [reply torch/csrc/jit/ivalue.h]: Yeah that is right, but you'll get the move constructor called twice instead of once. I guess it's not a big deal if we only hold a pImpl.
